pkg/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,82 @@
+package ratelimit
+
+import (
+	"time"
+
+	"go.uber.org/ratelimit"
+)
+
+type Event struct {
+	Timestamp time.Time
+	Tag       int
+}
+
+type RateLimiter struct {
+	outputCh     chan Event
+	src          <-chan Event
+	rt           ratelimit.Limiter
+	ch           chan Event
+	doneSender   chan struct{}
+	doneReceiver chan struct{}
+}
+
+func NewWithEPS(eventsPerSecond int64, src <-chan Event) *RateLimiter {
+	rt := newCommon(src)
+	rt.rt = ratelimit.New(int(eventsPerSecond))
+	return rt
+}
+
+func NewUnlimited(src <-chan Event) *RateLimiter {
+	rt := newCommon(src)
+	rt.rt = ratelimit.NewUnlimited()
+	return rt
+}
+
+func newCommon(src <-chan Event) *RateLimiter {
+	return &RateLimiter{
+		src:          src,
+		ch:           make(chan Event),
+		outputCh:     make(chan Event),
+		doneSender:   make(chan struct{}),
+		doneReceiver: make(chan struct{}),
+	}
+}
+
+func (rt *RateLimiter) Run() {
+
+	sender := func() {
```

Comment:
I'm a bit surprised we need two helper goroutines.
Additionally, we seem to add some overhead in the unlimited usecase. Do we have any option to optimize the unlimited use case so we don't introduce overhead (or as low as possible?)

Reply:
I have two goroutines because I want to decouple input channel from output channel.
I want to give the client the oportunity to always be able to write into the input channel with almost no bloking time while writing in the output channel just following the marked time period, so that's why I have one routine to be always reading from "src" and trying to write into "ch" but not stopping if "ch" already have an item, and the sender one reads from "ch" and send to output...

Reply:
circling back to this, decoupling input and output is fine, I'm just not sure we need two goroutines to achieve this; this may very well be the case, and so it needs a good description/rationale (code comments and/or commit message).
It may also be the case that the API we choose (exposing a channel) drives the implementation, e.g. making the two-goroutine approach the best; this is fine, provided that it is documented :)